### PR TITLE
Fix/midfielder positions

### PR DESCRIPTION
**Pull request by JAndringa**
### Comments for the reviewer
Some plays rely heavily on the use of enemy robots. In the rare case that these robots are absent the AI would do random stuff and sometimes even crash. This PR adds an alternative for these cases so that we do not draw against a team without robots.

###### Code Quality
- [x] Is the code is understandable and easy to read
- [x] Changes to the code comply with set clang-format rules
- [x] No use of manual memory control (e.g new/malloc/colloc etc)
- [x] Are (only) smart pointers used?

###### Testing
- [x] All tests are passing.
- [ ] I _added new / changed existing_ tests to reflect code changes (state why not otherwise!)
- [x] I tested my changes manually (Describe how, to what extent etc.)

###### Commit Messages
- [x] Commit message is saying what has been changed, **why** it was changed? Remember other developers might not know
  what the problem you are fixing was. Note also negative _decision_ (e.g., why did you not do particular thing)
  **TLDR: Commit message are comprehensive**
- [x] Commit messages follows the rules of https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: roboteam_ai/src/stp/plays/defensive/DefendPass.cpp [hunk 1]
```diff
@@ -115,15 +119,18 @@ void DefendPass::calculateInfoForDefenders() noexcept {
...
 
-    for (int i = 1; i <= 3; i++) {
-        if (!enemyMap.empty()) {
-            stpInfos["midfielder_" + std::to_string(i)].setPositionToDefend(enemyMap.rbegin()->second);
-            stpInfos["midfielder_" + std::to_string(i)].setBlockDistance(BlockDistance::HALFWAY);
-            enemyMap.erase(prev(enemyMap.end()));
-        } else {
-            break;
-        }
+    if (enemyMap.size() < 2) {
```
Comment (emielsteerneman) [inline roboteam_ai/src/stp/plays/defensive/DefendPass.cpp:122]: What happens if there is exactly one enemy? Is that enemy then not defended? Or is that enemy by default then the keeper and does not need defending? I'm not sure having a keeper is required

File: roboteam_ai/src/stp/plays/defensive/DefendPass.cpp [hunk 2]
```diff
@@ -137,9 +144,14 @@ void DefendPass::calculateInfoForRobotDefenders() noexcept {
     auto enemyRobots = world->getWorld()->getThem();
     auto enemyClosestToBall = world->getWorld()->getRobotClosestToBall(world::them);
 
+    if (!enemyClosestToBall.has_value()) {
+        roles.at(8) = std::make_unique<role::BallDefender>(role::BallDefender("robot_defender"));
```
Comment (emielsteerneman) [inline roboteam_ai/src/stp/plays/defensive/DefendPass.cpp:148]: Is there a possibility to not have the `8` hardcoded?

File: roboteam_ai/src/stp/plays/offensive/Attack.cpp [hunk 3]
```diff
@@ -100,20 +100,21 @@ void Attack::calculateInfoForRoles() noexcept {
         enemyMap.insert({score, enemy->getPos()});
     }
 
-    constexpr auto midfielderNames = std::array{"pass_defender_1", "pass_defender_2", "pass_defender_3"};
-    auto activeMidfielderNames = std::vector<std::string>{};
-    for (auto name : midfielderNames) {
-        if (stpInfos[name].getRobot().has_value()) activeMidfielderNames.emplace_back(name);
+    if (enemyMap.size() < 3) {
```
Comment (emielsteerneman) [inline roboteam_ai/src/stp/plays/offensive/Attack.cpp:103]: What if there are two opponents in the field? Are these left undefended?

**Review by emielsteerneman — CHANGES_REQUESTED** [review]
I am not actually sure how to test this play, since I don't know how to start a simulator with only us having robots in the field. Do you know how I can get that working?

File: roboteam_ai/src/stp/plays/defensive/DefendPass.cpp (hunk 1 above)
Comment (JAndringa) [reply roboteam_ai/src/stp/plays/defensive/DefendPass.cpp:122]: If there is one enemy they only have one robot that has a chance at scoring (enemy keeper is not included in this map). In that case our keeper should be enough to block the ball. A keeper is not required but its ID is set by the referee and can only be changed through the referee.

File: roboteam_ai/src/stp/plays/defensive/DefendPass.cpp (hunk 2 above)
Comment (JAndringa) [reply roboteam_ai/src/stp/plays/defensive/DefendPass.cpp:148]: Good one I should be able to change that

File: roboteam_ai/src/stp/plays/offensive/Attack.cpp (hunk 3 above)
Comment (JAndringa) [reply roboteam_ai/src/stp/plays/offensive/Attack.cpp:103]: This is a play in which we have ball possession, so if the opponents only have 2 robots it should be fine to leave them undefended as they will not be able to shoot unless they take the ball.

**Comment by JAndringa** [comment]
> I am not actually sure how to test this play, since I don't know how to start a simulator with only us having robots in the field. Do you know how I can get that working?

you can test this in gr sim by putting the opponents robots off and out of the field

File: roboteam_ai/src/stp/plays/defensive/DefendPass.cpp (hunk 1 above)
Comment (JAndringa) [reply roboteam_ai/src/stp/plays/defensive/DefendPass.cpp:122]: there are some defenders to block the ball indeed

**Review by emielsteerneman — APPROVED** [review]

**Pull request merged**